### PR TITLE
Use a shared enum for key prefixes

### PR DESCRIPTION
**Pull request by jacek-casper**
I need a way to query global state by prefixes for https://app.zenhub.com/workspaces/core-protocol-60953fafb1945f0011a3592d/issues/gh/casper-network/casper-node/4686. The existing code manually encodes byte arrays wherever a lookup by prefix was needed, but that's difficult to keep track of and it's pretty prone to errors, so I decided to create an enum that represents all possible key prefix combinations (excluding ones that are too broad, like querying all system/account/contract entities). This enum allows us to have a single data type for all the prefix lookups. Its deserializer is tested against the Key serializer to ensure that it remains binary compatible with the `Key` type. 

This enum can also be a useful way to expose the capability of querying the GS by a prefix to binary port consumers in a safe way, because we can prevent them from making very large queries (like querying with just a key tag) and we can restrict which `KeyPrefix` variants we allow in the binary port.

### CODE REVIEW

File: types/src/key_prefix.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    MessagesByEntity(EntityAddr),
+    /// Retrieves all messages for a given entity and topic.
+    MessagesByEntityAndTopic(EntityAddr, TopicNameHash),
+    /// Retrieves all named keys for a given entity.
+    NamedKeysByEntity(EntityAddr),
+    /// Retrieves all gas balance holds for a given purse.
+    GasBalanceHoldsByPurse(URefAddr),
+    /// Retrieves all processing balance holds for a given purse.
+    ProcessingBalanceHoldsByPurse(URefAddr),
+}
```
Comment (EdHastingsCasperAssociation) [inline types/src/key_prefix.rs]: i don't think this should be put in / published with `casper_types`; it just becomes yet another leaked implementation detail that causes breaking changes when we extend it / change it / go with a different implementation.

I would be more ok with it being put in `casper_storage`.

File: types/src/system/auction/bid_addr.rs [hunk 2]
```diff
@@ -135,13 +133,8 @@ impl BidAddr {
...
     /// Returns the common prefix of all delegators to the cited validator.
-    pub fn delegators_prefix(&self) -> Result<Vec<u8>, Error> {
-        let validator = self.validator_account_hash();
-        let mut ret = Vec::with_capacity(validator.serialized_length() + 2);
-        ret.push(KeyTag::BidAddr as u8);
-        ret.push(BidAddrTag::Delegator as u8);
-        validator.write_bytes(&mut ret)?;
-        Ok(ret)
+    pub fn delegators_prefix(&self) -> KeyPrefix {
+        KeyPrefix::DelegatorBidAddrsByValidator(self.validator_account_hash())
```
Comment (EdHastingsCasperAssociation) [inline types/src/system/auction/bid_addr.rs]: I don't agree with this...later logic could map this into an enum or whatever but the fundamental behavior should be as primitive as possible (in this case, the original implementation).

**Review by EdHastingsCasperAssociation — CHANGES_REQUESTED** [review]
See comments

File: types/src/key_prefix.rs (hunk 1 above)
Comment (jacek-casper) [reply types/src/key_prefix.rs]: I've added 2 versions of the enum, one in `storage` and one in `binary_port` and removed this one

File: types/src/system/auction/bid_addr.rs (hunk 2 above)
Comment (jacek-casper) [reply types/src/system/auction/bid_addr.rs]: I've reverted the code changes related to auctions

**Review by EdHastingsCasperAssociation — APPROVED** [review]

**Pull request merged**